src/iqm/qiskit_iqm/iqm_transpilation.py

```diff
@@ -0,0 +1,97 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transpilation tool to optimize the decomposition of 1 qubit gates tailored to IQM hardware."""
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.circuit import Qubit
+from qiskit.circuit.library import RGate
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes import Optimize1qGatesDecomposition, Unroller
+from qiskit.transpiler.passmanager import PassManager
+
+
+class IQMOptimize1QbDecomposition(TransformationPass):
+    r"""Optimize the decomposition of 1 qubit gates for the IQM gate set.
+
+    This optimisation pass expects the circuit to be correctly layouted and translated to the IQM architecture
+    and raises an error otherwise.
+    The optimisation logic follows the following steps:
+
+    1. Convert single qubit gates to :math:`U` gates and combine all neighbouring :math:`U` gates.
+    2. Convert :math:`U` gates according to
+       :math:`U(\theta , \phi , \lambda) ~ RZ(\phi + \lambda) R(\theta, \pi / 2  - \lambda)`.
+    3. Commute `RZ` gates to the end of the circuit using the fact that `RZ` and `CZ` gates commute, and
+       :math:`R(\theta , \phi) RZ(\lambda) = RZ(\lambda) R(\theta, \phi - \lambda)`.
+    4. Drop `RZ` gates immediately before measurements, and otherwise replace them according to
+       :math:`RZ(\lambda) = R(\pi, \lambda / 2) R(- \pi, 0)`.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._basis = ['r', 'cz']
+        self._intermediate_basis = ['u', 'cz']
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        self._validate_ops(dag)
+        # accumulated RZ angles for each qubit, from the beginning of the circuit to the current gate
+        rz_angles: list[float] = [0] * dag.num_qubits()
+        # convert all gates in the circuit to U and CZ gates
+        dag = Unroller(self._intermediate_basis).run(dag)
+        # combine all sequential U gates into one
+        dag = Optimize1qGatesDecomposition(self._intermediate_basis).run(dag)
+        last_op_is_measurement: dict[Qubit, bool] = {qubit: False for qubit in dag.qubits}
+        for node in dag.topological_op_nodes():
+            if node.name == 'u':
+                qubit_index = dag.find_bit(node.qargs[0])[0]
+                dag.substitute_node(
+                    node, RGate(node.op.params[0], np.pi / 2 - node.op.params[2] - rz_angles[qubit_index])
+                )
+                phase = node.op.params[1] + node.op.params[2]
+                dag.global_phase += phase / 2
+                rz_angles[qubit_index] += phase
+            if node.name == 'measure':
+                for qubit in node.qargs:
+                    last_op_is_measurement[qubit] = True
```

Comment:
Thinking out loud... This line is technically going to be triggered for mid-circuit measurements as well, and I think it is going to result into incorrect transpilation. Of course we do not yet support mid-circuit measurements, so if an execution is attempted for such a circuit, it is going to fail, but we should not rely on that, i.e. we should not say "the execution is going to fail anyways, so doesn't matter if we do the transpilation incorrectly". Transpiling can also be done independent from execution, so it should always produce correct results. I think we can choose one of the following options:

1. Raise error in this transpile pass if mid-circuit measurement is present.
2. Actually handle mid-circuit measurements correctly. It seems to me that this can easily: instead of keeping track of which qubit has a measurement, you can just reset the accumulated rz angle to zero every time a measurement is encountered for a qubit. At the end check which qubits have non zero rz angles, and add the two RGates at the end of the dag.

Any thoughts?

Reply:
2. works, too. But I think the current version will work as well because in line 70 `last_op_is_measurement` is set to `False` again if there is another operation on that qubit.

Reply:
Yeah, but the accumulated rz angle is not reset, right? So it will continue accumulating after the measurement. Mathematically this should not result into a wrong circuit, but we can use this chance to drop some rz

Reply:
I tested this numerically, and it seems both approaches (keeping or dropping the rz angle at mid-circuit measurement) provide the same probabilities for the measurement results if everything in the circuit is ideal, so both should be ok. Once we have mid-circuit measurements we could check which option works better with measurements that aren't perfectly ideal.

Reply:
Now we reset the rz angle at each measurement. Actually simplifies the code.